webhook-appbuilder-tutorial (#385)
AdobeDocs/commerce-extensibility: src/pages/webhooks/tutorial/index.md

With Adobe App Builder, you can build scalable, event-driven integrations using webhooks and runtime actions. These integrations allow you to extend Adobe Commerce capabilities without modifying the core codebase, making it easier to maintain and scale your applications. For example, you can use App Builder to validate data, such as product names, in real time through synchronous webhook integrations in Adobe Commerce as a Cloud Service (SaaS).

This tutorial provides instructions for setting up an App Builder project. It covers writing simple action code, configuring webhook subscriptions in Adobe Commerce, and debugging App Builder code using a debugger.

## How it works

This tutorial demonstrates how to extend Adobe Commerce using webhooks and Adobe App Builder. The integration allows you to validate product names in real time when products are added or updated in Adobe Commerce. The following provides an overview of how this example works:

1. Adobe Commerce triggers a webhook when specific events occur, such as when a new product is added or updated.

1. This webhook sends the relevant event payload (the product name, in this case) to an Adobe App Builder runtime action.

> 1. The App Builder action processes the payload and applies custom business logic. Validate that the product name does not include restricted terms like "test" or "example". Based on the logic, the action returns either a success or an error message.

Reviewer: ```suggestion
1. The App Builder action processes the payload and applies custom business logic and validates that the product name does not include restricted terms like "test" or "example". Based on the logic, the action returns either a success or an error message.
```

1. Adobe Commerce uses this response to determine whether to save changes to the product.

> This setup allows you to offload validation and custom logic from the Commerce codebase to the scalable, serverless infrastructure of Adobe App Builder.

Reviewer: ```suggestion
This example allows you to offload validation and custom logic from the Commerce codebase to the scalable, serverless infrastructure of Adobe App Builder.
```

## Set up the Adobe Developer Console and App Builder project locally

Before you can start building your App Builder application, you need to set up your development environment and create a project in the Adobe Developer Console. This process involves installing the necessary tools, configuring your workspace, and creating a new project that will host your App Builder application.

### Prerequisites

* An [Adobe Commerce as a Cloud Service](https://experienceleague.adobe.com/en/docs/commerce/cloud-service/overview) instance
* Access to [Adobe Developer Console](https://developer.adobe.com/console)
* Access to [Adobe Developer App Builder](https://developer.adobe.com/app-builder/docs/get_started/app_builder_get_started/set-up#access-and-credentials)

### Accessing the Developer Console

Ensure you have Developer access to the Adobe Developer Console. This is required to create and manage projects.

If you do not have developer access, your role displays as "User" in the top-right corner, and you will see a restricted access message on the Developer Console home page.

To proceed, request Developer Access from your organization admin.

> The following images indicate both restricted and developer access views.

Reviewer: ```suggestion
The following images contrast what you will see with and without developer access.
```

![With Developer Access:](../../_images/webhooks/tutorial/developer-access-dev-console.png)

### Create a new project in Adobe Developer Console

Follow the steps described in [Create an App Builder project](https://developer.adobe.com/commerce/extensibility/events/project-setup/). These steps are essential because the Adobe Developer Console provides the credentials and configuration required to deploy your App Builder app and access Adobe services like I/O Runtime and Commerce APIs. Without completing these steps, your app will not be able to authenticate or run within the Adobe ecosystem.

This procedure allows you to quickly set up a new App Builder project with the necessary configurations and dependencies. You must install multiple APIs in your project, though the exact APIs vary, depending on whether you have Adobe Commerce as a Cloud Service (SaaS) or Adobe Commerce on Cloud or on-premises (PaaS).

&#8203;<Edition name="saas" /> Install the following APIs in your **Stage** workspace:

*  Adobe Commerce as a Cloud Service API
*  I/O Events
*  Adobe I/O Events for Adobe Commerce

> &#8203;<Edition name="paas" /> Install the following APIs in your **Stage** workspace:

Reviewer: It doesn't really matter that they use the stage workspace, just as long as they use the same workspace throughout, right?

1. Run the `aio app use <path-to-json-config-file>` command to use the downloaded JSON file for your project.

   ```terminal
   You are currently in:
   1. Org: <org name>
   2. Project: <project name specified in development console>
   3. Workspace: Stage
   ```

1. Run the following command to initialize your project:

   ```bash
   aio app init <your project name>
   ```

   The command displays multiple prompts to configure your App Builder project. Follow these steps:
   * Select the correct **Organization** and **Project**.
   * Choose a template listed under **Supported by My Org** to ensure compatibility with your environment.
   * For the following prompts, make the corresponding selection:
     * **Select a template** - **All Templates → @adobe/generator-app-events-generic**.
     * **Which Adobe I/O App features do you want to use?** - **Actions: Deploy Runtime actions for I/O App features**
     * **Which type of actions do you want to generate?** - **Generic**.
     * **Which UI framework do you want to use?** - **React Spectrum 3**.

>      * **Provide a name for the action** -  Enter your preferred action name, such as `testwebhook`.

Reviewer: ```suggestion
     * **Provide a name for the action** - Enter your preferred action name, such as `testwebhook`.
```

   ```js
   const { Core, Events } = require('@adobe/aio-sdk') // Adobe I/O SDK modules
   const { errorResponse, stringParameters, checkMissingRequestInputs } = require('../utils') // Utility functions
   // 
   Main function executed by Adobe I/O Runtime
   async function main(params) {
    // Create a logger instance
    const logger = Core.Logger('main', { level: params.LOG_LEVEL || 'info' })
    logger.info('Calling the main action ')
    
    try {
      const response = { statusCode: 200 }
      
      // Check if product name contains 'test'; return error response if true
      if (/test/.test(params.product.name.toLowerCase())) {
      response.body = JSON.stringify({
        op: "exception",
        message: "Invalid product name >> " + params.product.name
      })
    } else {
      // Success response
      response.body = JSON.stringify({
        op: "success"
      })
    }
    
    return response
    } catch (error) {
      // Log error and return a 500 server error response
      logger.error(error)
      return errorResponse(500, 'server error ' + JSON.stringify(params), logger)
      }
    }
    exports.main = main
    ```

> 1. Copy the contents of the `utils.js` file provided below into the `utils.js` file located under the `actions` folder. The `utils.js` file  provides common utility functions used by actions. One of its main functions, `errorResponse`, helps create a consistent error response and optionally logs the error details. This utility is useful when validating input or handling failures in actions.

Reviewer: ```suggestion
1. Copy the contents of the `utils.js` file provided below into the `utils.js` file located under the `actions` folder. The `utils.js` file provides common utility functions used by actions. One of its main functions, `errorResponse`, helps create a consistent error response and optionally logs the error details. This utility is useful when validating input or handling failures in actions.
```

    *        e.g. 'missing xyz parameter'
    * @param {*} [logger] an optional logger instance object with an `info` method
    *        e.g. `new require('@adobe/aio-sdk').Core.Logger('name')`
    *
    * @returns {object} the error object, ready to be returned from the action main's function.
    *
    */
    function errorResponse (statusCode, message, logger) {
        if (logger && typeof logger.info === 'function') {
              logger.info(`${statusCode}: ${message}`)
            }
        return {
          error: {
            statusCode,
          body: {
            error: message
            }
          }
        }
      }
      
    module.exports = {
      errorResponse
    }
   ```

### Configure the `app.config.yaml` file

The `app.config.yaml` file, located in the root of your project, configures your Adobe App Builder project. It defines project metadata, runtime actions, web assets, and settings like which file to use as the entry point for each action.

Change the value of the `function` line from `index.js` reference to the location of your `validateProductName.js` file.

<InlineAlert variant="info" slots="text"/>

> You must specify `web: 'yes'` in the config. This ensures the action is exposed via a public URL and can be properly invoked by Adobe Commerce.

Reviewer: ```suggestion
You must specify `web: 'yes'` in the config. This ensures the action is exposed through a public URL and can be properly invoked by Adobe Commerce.
```

        actions:
          webhook:
            function: actions/testwebhook/validateProductName.js
            web: 'yes'
            runtime: nodejs:22
            inputs:
              LOG_LEVEL: info
            annotations:
              require-adobe-auth: false
              final: true
```
  
### Deploy and test

Run the following commands in your project directory:

1. Install all project dependencies:

   ```bash
   npm install

1. Build the web assets:

   ```bash
   aio app deploy
   ```
  
   The deploy command builds the application, deploys the action, and uploads web assets to the CDN. The output is similar to the following, showing the deployed action URL and the URL to access your application.

   ```terminal
   ✔ Built 2 action(s) for 'application'
   ✔ Building web assets for 'application'
   ✔ Deployed 1 action(s) for 'application'
   ✔ Deploying web assets for 'application'
   ✔ All static assets for the App Builder application in workspace: application were successfully deployed to the CDN. Files deployed :
   
   * 2 HTML page(s)
   * 2 Javascript file(s)
   * 3 .map file(s)
   * 1 CSS file(s)
   
   Your deployed actions:
   
   web actions:

>    -> https://1244026-appbuilderforextens-stage.adobeio-static.net/api/v1/web/appbuilderforextensibility/testwebhook
>    
>    To view your deployed application:
>    -> https://1244026-appbuilderforextens-stage.adobeio-static.net/index.html
>    
>    To view your deployed application in the Experience Cloud shell:
>    
>    -> https://experience.adobe.com/?devMode=true#/custom-apps/?localDevUrl=https://1244026-appbuilderforextens-stage.adobeio-static.net/index.html

Reviewer: ```suggestion
   -> https://1234567-appbuilderforextens-stage.adobeio-static.net/api/v1/web/appbuilderforextensibility/testwebhook
   
   To view your deployed application:
   -> https://1234567-appbuilderforextens-stage.adobeio-static.net/index.html
   
   To view your deployed application in the Experience Cloud shell:
   
   -> https://experience.adobe.com/?devMode=true#/custom-apps/?localDevUrl=https://1234567-appbuilderforextens-stage.adobeio-static.net/index.html
```

   
   skipping publish phase...
   Successful deployment
   ```

Make a note of the Web Action URL. You will need to specify it in the next step within the Commerce configuration.

#### Configure a webhook in the Admin

In the Commerce Admin, navigate to **System** > **Webhooks** > **Webhook Subscriptions** to display the Webhooks grid page. Click the **Add New Webhook** button.

1. Define the webhook settings as follows. The batch name and hook name can be any value. You can leave unlisted fields blank.

   Field | Value
   --- | ---
   **Webhook Method** | `observer.catalog_product_save_after`
   **Webhook Type** | `after`
   **Batch Name** | `checkproductName`
   **Hook Name** | `checkproductName`

>    **URL** | The URL you noted in the previous step, such as `https://1244026-appbuilderforextens-stage.adobeio-static.net/api/v1/web/appbuilderforextensibility/testwebhook`

Reviewer: ```suggestion
   **URL** | The URL you noted in the previous step, such as `https://1234567-appbuilderforextens-stage.adobeio-static.net/api/v1/web/appbuilderforextensibility/testwebhook`
```

   -> https://1244026-appbuilderforextens-stage.adobeio-static.net/index.html
   
   To view your deployed application in the Experience Cloud shell:
   
   -> https://experience.adobe.com/?devMode=true#/custom-apps/?localDevUrl=https://1244026-appbuilderforextens-stage.adobeio-static.net/index.html
   
   skipping publish phase...
   Successful deployment 🏄
   ```

Make a note of the Web Action URL. You will need to specify it in the next step within the Commerce configuration.

#### Configure the webhook in the Admin

In the Admin, navigate to **System**> **Webhooks** > **Webhook Subscriptions** to display the Webhooks grid page. Click the **Add New Webhook** button.

1. Define the webhook settings as follows. The batch name and hook name can be any value. You can leave unlisted fields blank.

   Field | Value
   --- | ---
   **Webhook Method** | `observer.catalog_product_save_after`
   **Webhook Type** | `after`
   **Batch Name** | `checkproductName`
   **Hook Name** | `checkproductName`
   **URL** | The URL you noted in the previous step, such as `https://1244026-appbuilderforextens-stage.adobeio-static.net/api/v1/web/appbuilderforextensibility/testwebhook`

> 1. Define a hook field to specify the payload you want to send to App Builder. Click the **Add Hook Field** button and fill in the following fields:

Reviewer: ```suggestion
1. Define a hook field to specify the payload you want to send to App Builder. Click the **Add Hook Field** button and enter the following information:
```

   Field | Value
   --- | ---
   **Name** | `product`
   **Source** | `data.product`
   **Active** | `Yes`
  

> Fill in the fields as shown in the screenshot below. You can choose any values for the Hook Name and Batch Name, but ensure **Hook Fields** match the screenshot exactly. In the URLfield, enter the Web Action URL you noted in the previous step.Save the Webhook.
> 
> ![WebHook Configuration](../../_images/webhooks/tutorial/webhook-config.png)
> 
> Below, you'll find the hook fields section. Here, you can specify the payload fields you want to send to App Builder.
> 
> ![WebHook Configuration](../../_images/webhooks/tutorial/hookfields-config.png)
> 

Reviewer: ```suggestion
```

Delete. this information is contained in the previous steps.

#### Testing the integration using webhooks

You now have App Builder code set up to validate the product name received from Commerce. Whenever a product is added or edited in Commerce, the webhook triggers and synchronously invokes the App Builder code to perform the validation.

From the Commerce Admin panel, navigate to **Catalog** > **Add New Product**. Enter the product details such as name, SKU, and price. Set the name as "testproduct", then click **Save**.

Upon saving, the webhook triggers the App Builder code, which validates the product name and possibly returns an error message. The error **Invalid product name >>** will be displayed in the Commerce UI, confirming that the integration is working as expected.

> As per the App Builder code logic, if the product name does not contain the word `test`, the product should save successfully without triggering an error.

Reviewer: ```suggestion
As configured, if the product name does not contain the word `test`, the product should save successfully without triggering an error.
```
